### ISSUE TITLE
ninja test now runs google-benchmark tests?

### ISSUE THREAD

**Issue by behdad** [harfbuzz/harfbuzz#4153]
With benchmarking enabled, it looks like "ninja test" also runs "google-benchmark" tests. That's weird. A minor thing though.

**Comment by khaledhosny** [comment]
I this a subproject? I think it might be caused by the `yield: true` in:
https://github.com/harfbuzz/harfbuzz/blob/ea17c7a81a743d7e319da0ff4111bcf650d2011b/meson_options.txt#L26-L27

We probably should drop it, I don’t think we are interested in running tests for subprojects we build.

**Comment by behdad** [comment]
Yes, it's a subproject.

**Comment by khaledhosny** [comment]
The yield stuff is not related. The latest version of the google-benchmark wrap file (the subproject stuff) now include tests and provide no option to disable them. To skip them locally, you can run the tests with `meson test --no-suite google-benchmark` (which is awkward), but I don’t know if it can be made the default.

**Comment by behdad** [comment]
It's alright. Just a minor thing. Thanks for looking into it.

**Comment by eli-schwartz** [comment]
> To skip them locally, you can run the tests with `meson test --no-suite google-benchmark` (which is awkward), but I don’t know if it can be made the default.

You could also do `meson test --suite harfbuzz` to avoid tests from *any* subproject.

**Comment by behdad** [comment]
> While at it: is there a way to instruct meson to spawn the `slow` test suite earlier, to better parallelize?

That helped a little. Not much.

**Comment by eli-schwartz** [comment]
> While at it: is there a way to instruct meson to spawn the `slow` test suite earlier, to better parallelize?

https://mesonbuild.com/Reference-manual_functions.html#test_priority

> specifies the priority of a test. Tests with a higher priority are started before tests with a lower priority. The starting order of tests with identical priorities is implementation-defined. The default priority is 0, negative numbers are permitted.

**Comment by eli-schwartz** [comment]
> Running `meson test` takes 14s. Running `meson test --no-suite google-benchmark` takes 11.5s.
> 
> Is there any way to make this default, or still no luck?

https://mesonbuild.com/Reference-manual_functions.html#add_test_setup

you can use:

- exclude_suites: A list of test suites that should be excluded when using this setup. Suites specified in the --suite option to meson test will always run, overriding add_test_setup if necessary. Since 0.57.0
- is_default: Set whether this is the default test setup. If true, the setup will be used whenever meson test is run without the --setup option. Since 0.49.0

I still think just running your own project tests and skipping anything else makes the most sense, but fair enough, it requires passing a command-line option by default...

**Comment by behdad** [comment]
> (You did say it was a minor thing at the time, so I didn't pursue this line of thought. My apologies.)

Of course never *expecting* anything.

Tests were running slow so I thought I'll take another look. The `MALLOC_CHECK_=2` being *very* slow on macOS was the main culprit. Then I kept going.

Thanks again.

Also since I have your attention: `meson test` is slightly annoying compared to `ninja test`, because trying to interrupt the running using `ctrl+c` doesn't work as expected in `meson test`. It just interrupts one test. One has to hold `ctrl+c` down to eventually get meson's attention. With `ninja test`, you just need to ctrl+c once. This has been one reason I've been using `ninja test` instead.

**Comment by behdad** [comment]
> Also since I have your attention: `meson test` is slightly annoying compared to `ninja test`, because trying to interrupt the running using `ctrl+c` doesn't work as expected in `meson test`. It just interrupts one test. One has to hold `ctrl+c` down to eventually get meson's attention. With `ninja test`, you just need to ctrl+c once. This has been one reason I've been using `ninja test` instead.

I'm testing with meson 1.4.1. Apologies if this has already been reported / fixed.

**Comment by eli-schwartz** [comment]
Meson 0.57 added this change:

```
    mtest: allow quickly interrupting the test run
    
    The new behavior of interrupting the longest running test with Ctrl-C is useful
    when tests hang, but not when the run is completely broken for some reason.
    Psychology tells us that the user will compulsively spam Ctrl-C in this case,
    so exit if three Ctrl-C's are detected within a second.
```

I'm not sure what to do about this. It is apparently, clearly useful to at least some people to be able to CTRL+C once and cancel a currently broken test, so that they can see what's going on with the rest of the tests?

Nothing has changed since then, I think.

**Comment by behdad** [comment]
> ou can use:
> 
> * exclude_suites: A list of test suites that should be excluded when using this setup. Suites specified in the --suite option to meson test will always run, overriding add_test_setup if necessary. Since 0.57.0
> * is_default: Set whether this is the default test setup. If true, the setup will be used whenever meson test is run without the --setup option. Since 0.49.0

This does it for me:
```diff
diff --git a/meson.build b/meson.build
index f3e43b595..f00b711d0 100644
--- a/meson.build
+++ b/meson.build
@@ -1,5 +1,5 @@
 project('harfbuzz', 'c', 'cpp',
-  meson_version: '>= 0.55.0',
+  meson_version: '>= 0.57.0',
   version: '10.2.0',
   default_options: [
     'cpp_eh=none',          # Just to support msvc, we are passing -fno-exceptions also anyway
@@ -459,6 +459,11 @@ if not get_option('docs').disabled()
   subdir('docs')
 endif
 
+add_test_setup('default',
+               exclude_suites: ['google-benchmark'],
+               is_default: true)
+
+
 configure_file(output: 'config.h', configuration: conf)
 
 alias_target('lib', libharfbuzz)
```

Thank you. @khaledhosny Possible to require 0.57.0 or 0.59.0?

**Comment by behdad** [comment]
> I'm not sure what to do about this. It is apparently, clearly useful to at least some people to be able to CTRL+C once and cancel a currently broken test, so that they can see what's going on with the rest of the tests?

Interesting. Anyway, with your other suggestion, my needs are satisfied.

Thanks again. I really appreciate you lurking here. :)

**Comment by eli-schwartz** [comment]
If you're worried about bumping the minimum version, an option is to only provide the "good" testsuite behavior when a new enough version of meson is used. Simply wrap the add_test_setup call in

```meson
if meson.version().version_compare('>=0.57.0')
    add_test_setup(.....)
endif
```

**Comment by eli-schwartz** [comment]
(I assume you probably want pre-0.57 support for people running Debian oldstable, which ships with 0.56)

**Comment by eli-schwartz** [comment]
Oh, from the PR adding the "three times" behavior there is also this amusing lifehack:

> I like this behavior (it's the one the test suite I wrote for OpenGL uses, lol). I usually just use ctrl + \ to fully kill things, but I don't see any harm in having ctrl + c three times quickly kill it too.

**Comment by behdad** [comment]
> (I assume you probably want pre-0.57 support for people running Debian oldstable, which ships with 0.56)

I defer to @khaledhosny on these things. Thanks. Committed.

**Comment by khaledhosny** [comment]
> > (I assume you probably want pre-0.57 support for people running Debian oldstable, which ships with 0.56)
> 
> I defer to [@khaledhosny](https://github.com/khaledhosny) on these things. Thanks. Committed.

If we can get away with using the version check, then I don’t see a need to raise the required version just yet.

**Comment by khaledhosny** [comment]
> ```diff
> +add_test_setup('default',
> +               exclude_suites: ['google-benchmark'],
> +               is_default: true)
> ```

This is causing a build failure when building with glib submodule:

```
meson.build:468:2: ERROR: 'glib:default' is already set as default. is_default can be set to true only once
```

**Comment by eli-schwartz** [comment]
Ugh, this feels wrong and like a meson bug -- but also not one we can fix for existing versions packaged in Debian, obviously...

It has been previously reported: https://github.com/mesonbuild/meson/issues/12700

Glib worked around this but only in a newer wrap than the one you are using.

**Comment by khaledhosny** [comment]
`meson wrap update glib` gives me 2.82.4-1 which still has this issue.

**Comment by behdad** [comment]
I'm sorry Khaled. I should have sent through PR.

**Comment by behdad** [comment]
> I'm sorry Khaled. I should have sent through PR.

Should I revert? Or disable being default for now?

**Comment by behdad** [comment]
> This is causing a build failure when building with glib submodule:

As a stopgap, is there a way to set our `is_default` to true only if glib is not used as a submodule?

**Comment by eli-schwartz** [comment]
> `meson wrap update glib` gives me 2.82.4-1 which still has this issue.

It is upstream in glib 2.83 only, I think. That's the unstable branch. It may be possible to ask them to backport that to  the stable series.

> As a stopgap, is there a way to set our `is_default` to true only if glib is not used as a submodule?

Yes. After the glib dependency is searched, you can use

```meson
glib_dep.type_name() != 'internal'
```

to check whether glib was found as a subproject or as a system installed dependency.

**Comment by khaledhosny** [comment]
I think we can close this now.
